This PR creates an actor to draw labels using texture maps

devmessias opened the pull request:
- I've created  the actor and tests in different files just for help me to develop this PR. But latter I'll put everything inside of  actors.py and test_actors.py. 

To create pre-built texture atlas for FURY you just need to put the ttf files inside of `data/files` and run the following command
```bash
python -c 'from fury import text_tools;text_tools._create_fury_system_atlas_fonts()'
```


The example 
[docs/examples/viz_billboard_labels.py](https://github.com/fury-gl/fury/blob/c05b1c06e84be97a32abf84565359dd4dfda3c13/docs/examples/viz_billboard_labels.py)
shows almost all the behaviors available for this PR

Result with bitmap texture
![Screenshot_20210815_134435](https://user-images.githubusercontent.com/6979335/129485892-f8136c04-a86b-4a1a-8680-9053c7b4e1ad.png)

Result with SDF
![image](https://user-images.githubusercontent.com/6979335/135487491-dd5b9445-b492-49b2-be9c-27128ffd969c.png)

devmessias committed:
feature: fast text rendering using billboards and bitmap fonts

devmessias committed:
feature: fast text rendering using billboards and bitmap fonts

devmessias committed:
test: color by character

devmessias committed:
test: color by character

devmessias committed:
fix: shader code rotations

devmessias committed:
fix: shader code rotations

pep8speaks commented:
Hello @devmessias! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`fury/actor_text.py`](https://github.com/fury-gl/fury/blob/9ca46b47791f49431b5d646da35270973b2dd559/fury/actor_text.py):

> [Line 81:5](https://github.com/fury-gl/fury/blob/9ca46b47791f49431b5d646da35270973b2dd559/fury/actor_text.py#L81): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 83:80](https://github.com/fury-gl/fury/blob/9ca46b47791f49431b5d646da35270973b2dd559/fury/actor_text.py#L83): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 79 characters)
> [Line 97:20](https://github.com/fury-gl/fury/blob/9ca46b47791f49431b5d646da35270973b2dd559/fury/actor_text.py#L97): [W292](https://duckduckgo.com/?q=pep8%20W292) no newline at end of file

* In the file [`fury/tests/test_text.py`](https://github.com/fury-gl/fury/blob/9ca46b47791f49431b5d646da35270973b2dd559/fury/tests/test_text.py):

> [Line 10:9](https://github.com/fury-gl/fury/blob/9ca46b47791f49431b5d646da35270973b2dd559/fury/tests/test_text.py#L10): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 11:80](https://github.com/fury-gl/fury/blob/9ca46b47791f49431b5d646da35270973b2dd559/fury/tests/test_text.py#L11): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (87 > 79 characters)
> [Line 46:11](https://github.com/fury-gl/fury/blob/9ca46b47791f49431b5d646da35270973b2dd559/fury/tests/test_text.py#L46): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 81:1](https://github.com/fury-gl/fury/blob/9ca46b47791f49431b5d646da35270973b2dd559/fury/tests/test_text.py#L81): [W391](https://duckduckgo.com/?q=pep8%20W391) blank line at end of file

devmessias committed:
this exclude font.png from .gitignore

devmessias committed:
this exclude font.png from .gitignore

devmessias committed:
transparency

devmessias committed:
transparency

devmessias committed:
this set offset ratio as a parameter

In order to control the spacing between characters

devmessias committed:
this set offset ratio as a parameter

In order to control the spacing between characters

devmessias committed:
uses an 8-bit chanel for texture bitmap fonts

Suggestion made by @filipinascimento

devmessias committed:
uses an 8-bit chanel for texture bitmap fonts

Suggestion made by @filipinascimento

devmessias committed:
this allows to choose the billboard text alignment

devmessias committed:
this allows to choose the billboard text alignment

devmessias committed:
check if the char can be rendered

devmessias committed:
check if the char can be rendered

devmessias committed:
a simple example using the billboard labels

devmessias committed:
a simple example using the billboard labels

devmessias commented on fury/actor_text.py:
```diff
@@ -0,0 +1,101 @@
+from fury import text_tools
+
+
+def bitmap_labels(
```
I don't know if is a good name. 
Maybe 
`billboard_labels` is more appropriate

devmessias committed:
move billboard_label actor to actor.py

devmessias committed:
move billboard_label actor to actor.py

devmessias commented on fury/actor.py:
```diff
@@ -2740,3 +2742,92 @@ def callback(
     return sq_actor
+
+
+def bitmap_labels(
```
Maybe `billboard_labels` is more suitable for this actor. 
@skoudoro @filipinascimento @Garyfallidis

devmessias replied:
Just  `actor.labels`? I don't like `bitmap_labels` because in the future we can use the same actor to draw text using msdf which has more than one channel and is not a bitmap.

skoudoro replied:
I would call it `styled_text`.

skoudoro replied:
final decision meeting (09/30/2021): 
- The name will be `label_fast`
- we need to rename `label` to `vector_text`
- After a release, `label_fast` will become `label`

devmessias commented on fury/text_tools.py:
```diff
@@ -0,0 +1,178 @@
+import fury
```
In  the future we need create more stuff to deal with text rendering. Like signed distance fields  MDSF, Unicode characters etc. I think is better to have a specific to deal with text stuff. But I don't know if `text_tools.py` is good name for that.

devmessias commented on fury/text_tools.py:
```diff
@@ -0,0 +1,178 @@
+    return image_arr, char2pos
+
+
+def get_positions_labels_billboards(
```
This function is where python has the worst performance in this PR.  Maybe in the future we must to create a optimized version using Numba LLVM compiler or cython.

filipinascimento replied:
In my testes this is not that bad, texture atlases are going to be created if they are not available for the first time, we should indicate the user what is happening.

devmessias replied:
At the first time this will not create a texture atlas. The FreeMono Atlas is already inside of this PR. This will create a new texture atlas only if the user requested a atlas with a better resolution or a different font.

devmessias replied:
I've changed that. Now is not possible to create a new atlas direct from` billboard_labels` actor. Instead you are allowed only to pass the name of the font to be used, `font_name`. 

The user can see the fonts available calling `text_tools.list_available_fonts()`.  

To create a new atlas to be used in the `billboard_lablels` you must explicit call `create_atlas_font`  method

```python
font_path = '/home/my_user/.../my_font.ttf'
text_tools.create_atlas_font(
    'NameOfMyNewFont', font_path=font_path, font_size_res=7)
# this will create a directory /home/.fury/font_atlas/NameOfMyNewFont
```
After this,  `NameOfMyNewFont` will be available
```python
new_font_actor = actor.bitmap_labels(
      [centers[5]], ['font size and format'],
      font_name='NameOfMyNewFont',
      align='center', scales=.05, colors=colors[3])
 ```

devmessias committed:
fix texture atlas from fretype-py

PEP8, missing int casts etc
original source code available here
https://github.com/rougier/freetype-py/blob/09a664d23850a7ba8cc208443b1c4f96f21116c6/examples/texture_font.py

devmessias committed:
fix texture atlas from fretype-py

PEP8, missing int casts etc
original source code available here
https://github.com/rougier/freetype-py/blob/09a664d23850a7ba8cc208443b1c4f96f21116c6/examples/texture_font.py

devmessias committed:
fix: texture atlas for a generic number of chanels

devmessias committed:
fix: texture atlas for a generic number of chanels

devmessias committed:
allows non-monospaced fonts: change from Pillow to freetype-py

devmessias committed:
allows non-monospaced fonts: change from Pillow to freetype-py

devmessias committed:
docs: an example which show some behaviors from billboard_labels

devmessias committed:
docs: an example which show some behaviors from billboard_labels

devmessias committed:
refactor: freetype-py code organization

devmessias committed:
refactor: freetype-py code organization

devmessias committed:
freetype-py as a optional requirement

devmessias committed:
freetype height scalling and offset

Co-authored-by: Filipi Nascimento Silva <filipinascimento@gmail.com>

devmessias committed:
freetype height scalling and offset

Co-authored-by: Filipi Nascimento Silva <filipinascimento@gmail.com>

devmessias committed:
remove requirements file

devmessias committed:
remove requirements file

devmessias committed:
fix: more suitable arg names

devmessias committed:
fix: more suitable arg names

devmessias committed:
fix: width spacing between freetype chars

devmessias committed:
fix: width spacing between freetype chars

devmessias committed:
requirements.txt was deleted by a mistake

devmessias committed:
requirements.txt was deleted by a mistake

devmessias committed:
test: if freeype-py is not available,  skip the test

devmessias committed:
test: if freeype-py is not available,  skip the test

devmessias committed:
allows to choose the size of the  texture font atlas

devmessias committed:
allows to choose the size of the  texture font atlas

filipinascimento commented on fury/text_tools.py:
```diff
@@ -0,0 +1,589 @@
+
+    if font_size == 7 and font_path is None:
+        font_path = f'{fury.__path__[0]}/data/files/FreeMono'
+        image_arr = Image.open(font_path+'.bmp')
```
Why 7?  Is this going to be bundled with fury? In that case there is another way to provide data. Checkout the network example that shows a wiki network. How about generating this in the building stage? As if there are changes in future, we the atlas will be always updated.

filipinascimento commented on fury/actor.py:
```diff
@@ -2740,3 +2742,103 @@ def callback(
+        font_size=font_size, font_path=font_path,
+        atlas_size=atlas_size, show=False)
+    padding, labels_positions,\
+        uv, relative_sizes = text_tools.get_positions_labels_billboards(
```
Positions are not going to be the final positions of each  quad (no padding applied to them) vtk clipping will remove the object if the center of the text fall off the visible area. This can be fixed later via an Issue.

filipinascimento reviewed (changes requested):
There are a few suggestions here, may need to cover more code in the tests, it works on all systems. Improvements could be done later as Issues.

devmessias committed:
improves the font management system.

This implements behaviors that will facilitate how the user can store
and access different fonts.

The pre-built fonts for FURY will be stored on "data/files/font_atlas".
Meanwhile by default the  texture atlas created by the user will be
stored in the user folder `.fury/font_atlas/font_name`.

To list all the available fonts the user can call
`text_tools.list_available_fonts`

devmessias committed:
improves the font management system.

This implements behaviors that will facilitate how the user can store
and access different fonts.

The pre-built fonts for FURY will be stored on "data/files/font_atlas".
Meanwhile by default the  texture atlas created by the user will be
stored in the user folder `.fury/font_atlas/font_name`.

To list all the available fonts the user can call
`text_tools.list_available_fonts`

devmessias committed:
inconsolata bold as the default font

devmessias committed:
inconsolata bold as the default font

devmessias committed:
include font-atlas subdirs

devmessias committed:
fix: example labels  pep8

devmessias committed:
fix: example labels  pep8

devmessias committed:
this allows to use different chars than ASCII on the texture atlas

devmessias committed:
this allows to use different chars than ASCII on the texture atlas

devmessias committed:
after this, centers will be the final quad positions

This will avoid the vtk removing the label if the center is outside of the window

devmessias committed:
after this, centers will be the final quad positions

This will avoid the vtk removing the label if the center is outside of the window

devmessias committed:
feature: SDF font rendering using billboards

devmessias committed:
feature: SDF font rendering using billboards

devmessias committed:
feature: sdf text border and opacity

devmessias committed:
feature: sdf text border and opacity

devmessias committed:
billboard_labels will use the  vtkOpacity instead of a custom opacity uniform argument

devmessias committed:
billboard_labels will use the  vtkOpacity instead of a custom opacity uniform argument

skoudoro commented on docs/examples/viz_billboard_labels.py:
```diff
@@ -0,0 +1,121 @@
+"""
+=================================
+Billboard labels actor behaviosrs
```
typo

skoudoro commented on docs/examples/viz_billboard_labels.py:
```diff
@@ -0,0 +1,121 @@
+
+# Create a window FURY
+scene = window.Scene()
+# scene.background((1., 1., 1.))
```
to remove if not used

skoudoro commented on docs/examples/viz_billboard_labels.py:
```diff
@@ -0,0 +1,121 @@
+Billboard labels actor behaviosrs
+=================================
+
+This examples shows some functionalities of the `billboard_labels` actor.
```
- no `s` in `example`.
- actor name to update

skoudoro commented on docs/examples/viz_billboard_labels.py:
```diff
@@ -0,0 +1,121 @@
+and offsets.
+
+If you want to see the labels with a different font, you need to install
+the Freetype library and freetype py (`pip install freetype py`).
```
The sentence is not clear, is it 2 libraries?

skoudoro commented on docs/examples/viz_billboard_labels.py:
```diff
@@ -0,0 +1,121 @@
+    scales=.05, colors=colors[3])
+
+scene.add(offset_change_actor)
+###############################################################################
```
empty line above needed

skoudoro commented on docs/examples/viz_billboard_labels.py:
```diff
@@ -0,0 +1,121 @@
+if interactive:
+    window.show(scene, size=(600, 600), order_transparent=True)
+
+window.record(scene, out_path='viz_billboard_labels.png', size=(600, 600))
```
`out_path` name should be change in the near future

skoudoro commented on docs/examples/viz_billboard_labels_new_font.py:
```diff
@@ -0,0 +1,61 @@
+=================================
+
+If you want to see the labels with a different font, you need to install
+the Freetype library and freetype-py (`pip install freetype-py`).
```
same comment as above

skoudoro commented on docs/examples/viz_billboard_labels_new_font.py:
```diff
@@ -0,0 +1,61 @@
+import numpy as np
+
+# Set to True to enable user interaction
+interactive = True
```
False

skoudoro commented on docs/examples/viz_billboard_labels_new_font.py:
```diff
@@ -0,0 +1,61 @@
+    # We start chosing a path for the TTF file. Here we use the Roboto font
+    # that is available on the FURY examples folder.
+    font_path = font_path = f'{fury.__path__[0]}/data/files/RobotoMonoBold700.ttf'
+###############################################################################
```
need an empty line above

skoudoro commented on docs/examples/viz_billboard_labels_new_font.py:
```diff
@@ -0,0 +1,61 @@
+# To create a label with a different font, you need to have the FreeType
+# library and freetype-py installed.
+#
+if text_tools._FREETYPE_AVAILABLE:
```
Why is it protected (with an underscore)?

skoudoro replied:
the underscore is not needed

skoudoro commented on docs/examples/viz_billboard_labels_new_font.py:
```diff
@@ -0,0 +1,61 @@
+
+    # We start chosing a path for the TTF file. Here we use the Roboto font
+    # that is available on the FURY examples folder.
+    font_path = font_path = f'{fury.__path__[0]}/data/files/RobotoMonoBold700.ttf'
```
- Why twice `font_path`?
- `fury.__path__[0]` there is an existing variable instead of this I think. To check-in `fury.data` or `fetcher`

skoudoro commented on docs/examples/viz_billboard_labels_new_font.py:
```diff
@@ -0,0 +1,61 @@
+# Then we create the texture atlas for the font. The `font_size_res`
+# argument controls the quality of the font rendering, the higher the better
+#
+    text = 'A custom font with special characters like: ç, ã and à'
+    # # The `label` need to have special characters thus we will tell the
+    # # `create_atlas_font` to draw those characters.
+    chars = list(set(text))
+    text_tools.create_new_font(
+        'FreeMonoWithSpecial', font_path=font_path, font_size_res=10,
+        chars=chars, force_recreate=True)
+    new_font_actor = actor.bitmap_labels(
+        [np.array([0., 0., 0.])], [text],
+        font_name='FreeMonoWithSpecial',
+        align='center', scales=0.1,)
+
+    scene.add(new_font_actor)
```
why is it indented?

skoudoro commented on docs/examples/viz_billboard_labels_new_font.py:
```diff
@@ -0,0 +1,61 @@
+if interactive:
+    window.show(scene, size=(600, 600), order_transparent=True)
+
+window.record(scene, out_path='viz_billboard_labels.png', size=(600, 600))
```
same comment as above for `out_path`

skoudoro commented on docs/examples/viz_huge_amount_of_labels.py:
```diff
@@ -0,0 +1,116 @@
+"""
+============================================================
+This examples shows how to visualize a large amount of labels
```
no s to example

skoudoro commented on docs/examples/viz_huge_amount_of_labels.py:
```diff
@@ -0,0 +1,116 @@
+###############################################################################
+# Then let's download some available datasets.
+
+from fury.data.fetcher import fetch_viz_wiki_nw
```
`from fury.data` instead of `from fury.data.fetcher`

skoudoro commented on fury/actor.py:
```diff
@@ -2740,3 +2742,150 @@ def callback(
+        border_type='halo',
+        use_sdf=True,
+        ):
+    """Create a bitmap label actor that always faces the camera.
```
`bitmap text actor` instead

skoudoro commented on fury/actor.py:
```diff
@@ -2740,3 +2742,150 @@ def callback(
+
+    Parameters
+    ----------
+    centers : ndarray, shape (N, 3)
+    labels  : list
+        list of strings
+    colors : array or ndarray
+    scales : float, optional
+
+    align : str, {left, right, center}
+    x_offset_ratio : float
+        Percentage of the width to offset the labels on the x axis.
+    y_offset_ratio : float
+        Percentage of the height to offset the labels on the y axis.
+    font_size : int, optional
+        size of the text
+    font_name : str, optional
+        name of the font. A list of available fonts can be found in
+        `fury.text_tools.list_fonts_available()`.
+    use_sdf : bool, optional
+        If True, the labels will be rendered using a signed distance field.
```
missing some `optional` and some description

skoudoro commented on fury/shaders/text_billboard_impl.frag:
```diff
@@ -0,0 +1,27 @@
+// float s = 1;
+// float edgeWidth = 0.0;
+// edgeWidth = edgeWidth/2.;
+// float minSdf = 0.5/2.0;
+// vec2 d = abs(point.xy) - vec2(s, s);
+// float sdf = -length(max(d,0.0)) - min(max(d.x,d.y),0.0);
+
+// if (sdf<0.0) discard;
+
+vec4 rgba = vec4(color, vtkOpacity);
+// if (edgeWidthNew > 0.0){
+//     if (sdf < edgeWidthNew) {
+//         rgba = vec4(edgeColor, edgeOpacity);
+//     }
+// }
+
+    //vertexMC.xyz - center
+
+//texture(charactersTexture, UV);
+// if (texture(charactersTexture, UV).rgb != vec3(0.0, 0.0, 1.0)) discard;
+if (texture(charactersTexture, UV).r < 1.0) discard;
+// if (texture(charactersTexture, UV).r < 1.0) {
+//  rgba.rgb = vec3(1., 1.0, 1.0);
+// }
+//rgba.w = 0.0;
+
+fragOutput0 = rgba;
```
Can you remove all comments or add explanation otherwise, they bring confusion

skoudoro commented on fury/utils.py:
```diff
@@ -855,6 +855,35 @@ def rotate(actor, rotation=(90, 1, 0, 0)):
         prop3D.SetOrientation(newTransform.GetOrientation())
 
 
+def one_chanel_to_vtk(data):
```
missing a `n` in channel

skoudoro replied:
Also, I think this function should be in the `convert` module

skoudoro replied:
No sure I understand this function. Why do you mean by one channel? it seems you expect only a 2D image so there is only 1 channel. it would make more sense if you were expected a 3D image and the index of the channel that you want. Can you clarify ?

skoudoro replied:
Can you answer the questions above

skoudoro commented on fury/text_tools.py:
```diff
@@ -0,0 +1,844 @@
+_FONT_PATH_USER = pjoin(fury_home, 'font_atlas')
+
+
+def norm(img, vmin=0, vmax=255):
+    if vmin == 0:
+        img = np.interp(img, (img.min(), img.max()), (vmin, vmax))
+        # img /= np.max(img)
+        # img *= vmax
+        # img[img > 0.75] = 0.
+    elif vmin > 0:
+        # pass
+        img = np.interp(img, (img.min(), img.max()), (vmin, vmax))
+    img = img.astype(np.uint8)
+    return img
```
I am surprised that we do not have this function already somewhere or neither in `numpy`.  Can you check it? 
Also, this is an utility function, I recommend to move it somewhere else

devmessias replied:
I've created that because we can have different strategies to normalize the SDF.  This different strategies controls how the field will decay which directs impact the redering result. For example, 
```python
        img = np.interp(img, (img.min(), img.max()), (vmin, vmax))
```
or 
```python
img /= np.max(img)
img *= vmax
img[img > threshold] = 0.
 ```       
@filipinascimento can we discuss which behavior we will choose?

skoudoro commented on fury/text_tools.py:
```diff
@@ -0,0 +1,844 @@
+
+    Returns
+    -------
+    distComb : ndarray
```
`dist_comb` instead

skoudoro commented on fury/text_tools.py:
```diff
@@ -0,0 +1,844 @@
+    img_black_px[img > 0] = 0
+    img_black_px[img == 0] = 1
+    dist = ndimage.distance_transform_edt(img_white_px)[img > 0]
+    distNeg = ndimage.distance_transform_edt(img_black_px)[img == 0]
```
dist_neg

skoudoro commented on fury/text_tools.py:
```diff
@@ -0,0 +1,844 @@
+    distNeg = border_px - norm(distNeg, distNeg.min(), border_px)
+    # distNeg[distNeg< 255*0.25] = 0.
+    dist = norm(dist, border_px + 1, 255)
+    distComb = img.copy()
```
dist_comb

skoudoro commented on fury/text_tools.py:
```diff
@@ -0,0 +1,844 @@
+        The path to the font file.
+    folder : str
+        The path to the folder where the font will be saved.
+    font_size_res : int
```
optional

skoudoro reviewed (commented):
Hi @devmessias,

Thank you for this nice feature. I still need to go deeper but see below for some quick comments.
- Please, make sure to clean and remove all commented code
- Please, fix all pep8
- Strange location for `font.bmp`. I am not sure if it is needed

devmessias committed:
fix: pep8 and typos

devmessias committed:
fix: pep8 and typos

devmessias committed:
fix: trailing whitespace

devmessias committed:
fix: trailing whitespace

devmessias committed:
docs: improves the text of the billboad_actor examples

devmessias committed:
docs: improves the text of the billboad_actor examples

skoudoro commented:
Hi @devmessias,

there is a conflict, Can you rebase this PR? Thank you

devmessias commented:
Result with SDF
![image](https://user-images.githubusercontent.com/6979335/135486105-fef1528b-ede8-4b9b-949a-e4d518af54fc.png)

devmessias committed:
fix: vtk imports

devmessias committed:
change from bitmap_labels to billboard_labels

devmessias commented:
After df41a9e the actor is called using `billboard_labels` instead of `bitmap_labels`

devmessias committed:
fix : billboard_labels to label_fast

devmessias commented:
Hi all. I'll focus on this PR.

devmessias committed:
fix: this removes all unnecessary comments

In addition, comments used to perform string replacement in the shader compile phase will start with STR_REPLACEMENT after this commit

devmessias committed:
simplifies the func that normalizes the SDF glyph

    We choosed to use a simple normalization  with a linear
    transformation. Maybe there is a better way to do this.
    We should discuss that latter.

Garyfallidis commented:
@devmessias This needs a rebase / conflict resolution.

devmessias commented:
> @devmessias This needs a rebase / conflict resolution.

Ok. I'll do that.

skoudoro commented:
Hi @devmessias,

There is still some conflict. Any chance to have an update on this PR?

devmessias commented:
> Hi @devmessias,
> 
> There is still some conflict. Any chance to have an update on this PR?

Hi @skoudoro . I was working on the  #437   PR. @filipinascimento believes that is ready to be merged. Now I'm free to work on #489. I'll do the rebase today after lunch.

skoudoro commented:
ok thanks for the feedback, I will look at https://github.com/fury-gl/fury/pull/437 since it is ready

devmessias commented:
@skoudoro  ping. rebase done

devmessias committed:
tests: fix test_atlas

skoudoro commented on fury/tests/test_text.py:
```diff
@@ -0,0 +1,70 @@
+import fury
+
+
+def test_create_atlas():
+    if not text_tools._FREETYPE_AVAILABLE:
+        print('Bitmap text not tested (FREETYPE is not available)')
+        return
+
+    name = 'test_fonts_1238210930'
+    font_path_atlas = text_tools._FONT_PATH_USER
+    font_path = f'{fury.__path__[0]}/data/files/FreeMono.ttf'
+    text_tools.create_new_font(
+        name=name,
+        font_size_res=10, font_path=font_path, show=False)
+    fonts = text_tools.list_fonts_available(True)
+    shutil.rmtree(f'{font_path_atlas}/{name}')
+    if name not in fonts.keys():
+        raise FileNotFoundError(f'Font {name} was not created')
```
I do not see any tests.

What are you testing here?

skoudoro commented on fury/tests/test_text.py:
```diff
@@ -0,0 +1,70 @@
+
+    name = 'test_fonts_1238210930'
+    font_path_atlas = text_tools._FONT_PATH_USER
+    font_path = f'{fury.__path__[0]}/data/files/FreeMono.ttf'
```
same comment as above concerning the path

skoudoro commented on fury/tests/test_text.py:
```diff
@@ -0,0 +1,70 @@
+        raise FileNotFoundError(f'Font {name} was not created')
+
+
+def test_atlas():
+    if not text_tools._FREETYPE_AVAILABLE:
+        print('Bitmap text not tested (FREETYPE is not available)')
+        return
+
+    texture_atlas = text_tools.TextureAtlas(num_chanels=1)
+    font_path = f'{fury.__path__[0]}/data/files/FreeMono.ttf'
+    tf = text_tools.TextureFont(texture_atlas, font_path, 100)
+    tf.load('a sdf123')
```
same comment as above ? what is the test here ?

devmessias replied:
If the method responsible to create the atlas is working or not. Yeah, test_atlas is not a good name, I'll improve that

skoudoro replied:
I mean, your tests do not have any assertions. so not sure what you are really testing, they look like mini examples. 
The name is ok.

devmessias replied:
Ok, I get your point now. This only verifies that the file is written in the proper location and it's not corrupted. In both cases, the exceptions will be raised. But we can put a try, except and raise to become more meaningful

skoudoro commented on fury/tests/test_text.py:
```diff
@@ -0,0 +1,70 @@
+    tf.load('a sdf123')
+
+
+def test_bitmap_actor():
+    N = 10
+    colors = (0, 0.8, .5)
+    colors_spheres = colors
+
+    labels = ['AbrBac..ooo0123_///::adabra_ 1664123!@/?*...)(']
+    centers = np.random.normal(0, 10, size=(N, 3))
+    if N > 1:
+        min_s = 5
+        max_s = 10
+        chars = [chr(i) for i in range(32, 134)]
+        labels = [
+            ''.join(
+                    np.random.choice(
+                        chars,
+                        size=np.random.randint(min_s, max_s)
+                    )
+                )
+            for i in range(N)
+        ]
+        colors = []
+        colors_spheres = []
+        for label in labels:
+            c = np.random.uniform(0, 1, size=3)
+            colors_spheres.append(c)
+            for _ in label:
+                colors.append(c)
+    spheres = actor.markers(centers, colors_spheres, scales=.1)
+    my_text_actor = actor.label_fast(
+        centers, labels, colors=colors, scales=.1,
+        y_offset_ratio=1,
+        align='center', font_name='FreeMono')
+    showm = window.ShowManager(size=(700, 200))
+    showm.scene.add(my_text_actor)
+    showm.scene.add(spheres)
```
same comment as above, what is the test here ?

skoudoro commented on fury/text_tools.py:
```diff
@@ -0,0 +1,867 @@
+
+"""
+
+import sys
+import numpy as np
+from scipy import ndimage
+from PIL import Image
+import pickle
+import os
+from os.path import join as pjoin
```
pep8 import order

skoudoro commented on fury/text_tools.py:
```diff
@@ -0,0 +1,867 @@
+try:
+    import freetype as ft
+
+    _FREETYPE_AVAILABLE = True
+except ImportError:
+    _FREETYPE_AVAILABLE = False
```
You can remove the underscore here

skoudoro commented on fury/text_tools.py:
```diff
@@ -0,0 +1,867 @@
+import fury
+from fury.data.fetcher import fury_home
+
+_FONT_PATH_DEFAULT = f"{fury.__path__[0]}/data/files/font_atlas"
+_FONT_PATH_TTF = f"{fury.__path__[0]}/data/files/"
```
`from fury.data import DATA_DIR` and use it

skoudoro reviewed (commented):
Hi @devmessias,

Thank for the update. here are some comments:
- strange locations for `font.bmp`. Do you really need it?
- please, finish addressing the pepe8 
-  finalize all the tests
- the tutorials are really cool!

devmessias commented:
> 

Regarding font.bmp we need this stuff to avoid injecting the freetype-lb dependency. In that way, the freetype-lib becomes a requirement only for users who want to create new fonts using different font faces.

skoudoro commented:
> Regarding font.bmp we need this stuff to avoid injecting the freetype-lb dependency.

ok, so can you move it to `fury/data/files`

devmessias committed:
refactor: new FURY folder config

Fix pep8 order imports

skoudoro commented:
Hi @devmessias,

Streaming PR is merged, it would be good to finalize this one. Do you have any updates?

devmessias committed:
fix: vtk/fury imports

devmessias committed:
fix: uses the fury shader import

devmessias committed:
style: remove black fmt

devmessias committed:
fix: bring back Cython optional req

devmessias committed:
fix: this test is wrong

devmessias commented:
ping @filipinascimento  and @skoudoro . Rebase done! I've created an issue regarding all the failed tests in this PR https://github.com/fury-gl/fury/issues/705

skoudoro commented:
Thank you @devmessias, I was playing with your PR and yes it is ready to go. Any additional comments by other dev?

devmessias commented:
@filipinascimento  ping

filipinascimento reviewed (approved):
I have no more questions. I still need to understand why the SDF is too blocky. However let's do that in future additions. I think we should merge this asap, so that Helios can start using fury/master.

skoudoro commented:
All CI's are failing due to previous segfault which are fixed now. Can you rebase this PR @devmessias to make sure there is not any hidden errors?

Garyfallidis commented:
Ping @devmessias. A rebase is required.
